### DISCUSSION TITLE
Documentation request: Performance

### DISCUSSION THREAD

**Discussion by nielsbosma** [AvaloniaUI/Avalonia#8300] [Ideas]
Would be great with a list of things to consider when optimizing performance in Avalonia. 

I assume some of these also apply to Avalonia?:

https://dzone.com/articles/15-wpf-performance-tips-for-2019
http://www.wpftutorial.net/10performancetips.html

**Comment by timunie** [comment]
Hi, 

one thing that really can improve performance are `CompiledBindings` see. https://docs.avaloniaui.net/docs/data-binding/compiledbindings

moreover if you need to display a lot of data in a `DataGrid` or you have a `TreeView` with many nodes consider using `TreeDataGrid`: https://docs.avaloniaui.net/docs/controls/treedatagrid

Happy coding
Tim

**Reply by nielsbosma** [reply]
Does the TreeDataGrid support virtualization? or why is it faster?

**Reply by wieslawsoltes** [reply]
Yes it does and also very useful if you need virtualized tree as it has hierarchical data templates so you can use it instead of TreeView.

**Reply by nielsbosma** [reply]
I'm only using the DataGrid part but with thousands of rows with sometimes quite complex cells. Would it still be virtualized?

**Reply by timunie** [reply]
Yes it will be. Also normal DataGrid is virtualized, but TreeDataGrid was made from scratch and is even better in performance view.

**Reply by danwalmsley** [reply]
DataGrid is a terrible control, avoid at all costs... if you dont need editing, stick to TreeDataGrid.

**Comment by nielsbosma** [comment]
Use ItemsRepeater instead on ItemsControl?

ItemsRepeater supports virtualization.

**Reply by danwalmsley** [reply]
You can use ListBox to get virtualisation and turn off selection.

**Reply by maxkatz6** [reply]
That's complex one. Both solutions support virtualization. 
But we still actively use ItemsPresenter (used inside of ItemsControl) in our list controls because it's more stable. ItemsRepeater was ported from UWP and it has some issues when items have non-uniform height (or width for horizontal scrolling).
Answer would be to try both and see what's perform better and smoother for you.
